### PR TITLE
MSRunSample unique constraint code version of null == null

### PR DESCRIPTION
**Pull request by hepcat72**
## Summary Change Description

Changed the unique constraints in MSRunSample to ensure only a single placeholder record can exist for any one sequence and sample combo.

Details:

- Changed handle_load_db_errors to build ConflictingValueError exceptions when the unique constraint has a defined condition attribute
- Added method: find_conflicts
- Made handle_load_db_errors able to identify ValidationErrors associated with "Constraint <constraint name> is violated"
- Changed get_unique_constraint_fields to mix in Q expressions when a unique constraint has a defined condition.
- Changed the MSRunSample ordering to use ArchiveFile.filename instead of whatever the record evaluates to (the ID?)
- Updated the accucor loader to handle the new exceptions

Note, the MSRunSample ordering might need to be checked.

## Affected Issues/Pull Requests

- Partially addresses #947 (but in a different way)

Note, a follow-up PR will be created to migrate the data, so this should not be merged upon approval until that data migration is done.  I was thinking that I would create dummy accucor records for the MSRunSample records that run afoul of the new unique constraint, which, on production is comprised of about 7 compounds, about 4 accucor files and about 52 samples.

## Review Notes

See comments in-line.

## Checklist

This pull request will be merged once the following requirements are met.  The
author and/or reviewers should uncheck any unmet requirements:

- Review requirements
  - Minimum approvals: 1 
  - No changes requested
  - All blocking issues resolved by reviewers
  - Specific reviewers: @__add_username_here__
    
  - Review period: 2 days 
- Associated issue/pull request requirements:
  
  - [x] All requirements in affected issues marked "resolved" are satisfied
  - [x] All required pull requests are merged *(or none)*
- Basic requirements
  
  - [x] [All linters pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting)
  - [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All conflicts resolved
- Overhead requirements
  
  - [x] [New/changed method tests implemented/updated *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [ ] [Updated *Unreleased* section of `changelog.md` *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md)
  - [x] [Migrations created & committed *(or no change)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)

### CODE REVIEW

File: DataRepo/loaders/accucor_data_loader.py [hunk 1]
```diff
@@ -1679,7 +1680,16 @@ def load_data(self):
...
+                if not handle_load_db_errors(
+                    e,
+                    MSRunSample,
+                    msrunsample_dict,
+                    aes=self.aggregated_errors_object,
+                    conflicts_list=self.conflicting_msrun_samples,
+                    sheet="absolte" if self.isocorr_format else "Corrected",
+                    file=peak_annotation_file.filename,
+                ):
+                    self.aggregated_errors_object.buffer_error(e)
```
Comment (hepcat72) [inline DataRepo/loaders/accucor_data_loader.py:1692]: C. This turns the new `ValidationError`s associated with the code-based unique constraint violationsn into more useful `ConflictingValueError`s.

File: DataRepo/models/msrun_sample.py [hunk 2]
```diff
@@ -119,23 +120,20 @@ class MSRunSample(HierCachedModel, MaintainedModel):
     class Meta:
         verbose_name = "mass spectrometry run sample"
         verbose_name_plural = "mass spectrometry samples"
-        ordering = ["msrun_sequence", "sample", "ms_raw_file", "ms_data_file"]
+        ordering = [
+            "msrun_sequence",
+            "sample",
+            "ms_raw_file__filename",
+            "ms_data_file__filename",
```
Comment (hepcat72) [inline DataRepo/models/msrun_sample.py:127]: C. Minor unrelated change.

File: DataRepo/models/utilities.py [hunk 3]
```diff
@@ -371,43 +388,16 @@ def handle_load_db_errors(
...
+            if find_conflicts(
+                model,
+                rec_dict,
+                aes=aes,
+                conflicts_list=conflicts_list,
+                rownum=rownum,
+                sheet=sheet,
+                file=file,
+            ):
+                return True
```
Comment (hepcat72) [inline DataRepo/models/utilities.py:400]: C. The above code was moved into a new method below so it could be re-used in the new case of unique constraint violations arising from `full_clean` (a code-based check as opposed to a database check).

File: DataRepo/models/utilities.py [hunk 4]
```diff
@@ -468,5 +457,89 @@ def handle_load_db_errors(
...
+
+        # Retrieve the record with the conflicting value(s) that caused the unique constraint error using the
+        # unique fields
+        q = Q()
+        for uf in combo_fields:
+            if isinstance(uf, Q):
+                q &= uf
+            elif rec_dict[uf] is not None:
+                # We do not need to check None values.  None != None
+                q &= Q(**{f"{uf}__exact": rec_dict[uf]})
```
Comment (hepcat72) [inline DataRepo/models/utilities.py:529]: The code in this method is essentially the same as where it was moved from except for this bit, to handle the `Q` expression that comes from the code-based null=null unique constraint.

File: DataRepo/models/utilities.py [hunk 5]
```diff
@@ -429,8 +419,7 @@ def handle_load_db_errors(
                     missing_list.append(err)
                     return True
                 elif aes:
-                    for err in errs:
-                        aes.buffer_error(err)
+                    aes.buffer_error(err)
```
Comment (hepcat72) [inline DataRepo/models/utilities.py:422]: C. When I moved the code above, I was surprised `errs` was no longer defined.  It's a different block.  I'm too used to perl's lexical scoping.

File: DataRepo/tests/management/test_load_accucor_msruns.py [hunk 6]
```diff
@@ -466,40 +466,37 @@ def test_ambiguous_msruns_error(self):
             )
         # Check second file failed (duplicate compound)
         aes = ar.exception
-        print(f"{aes}")
         self.assertEqual(1, len(aes.exceptions))
         self.assertTrue(isinstance(aes.exceptions[0], AmbiguousMSRuns))
 
     @tag("multi-msrun")
-    def test_resolve_ambiguous_msruns_error(self):
+    def test_msrun_sample_null_equals_null_unique_constraint(self):
```
Comment (hepcat72) [inline DataRepo/tests/management/test_load_accucor_msruns.py:473]: C. Reversed this test from allowing multiple representations / placeholders, to ensuring it raises an error instead.

File: DataRepo/models/msrun_sample.py (hunk 2 above)
Comment (hepcat72) [reply DataRepo/models/msrun_sample.py:127]: Just tested [this page](http://127.0.0.1:8000/DataRepo/msrunsamples/) in my sandbox, after having loaded one of Michael's studies, and didn't have an issue, so I think this is good.

File: DataRepo/models/utilities.py (hunk 4 above)
Comment (lparsons) [reply DataRepo/models/utilities.py:529]: Thanks for noting this, it did make following this simpler.

File: DataRepo/models/msrun_sample.py (hunk 2 above)
Comment (lparsons) [reply DataRepo/models/msrun_sample.py:127]: Makes sense.

**Review by lparsons — APPROVED** [review]
Really nice solution Rob. Having the `utilities` methods be aware of the `ValidationError` and handling that in `find_conflicts` makes this a pretty seamless and elegant solution. 👍🏻
